### PR TITLE
Node generic 2

### PR DESCRIPTION
**Pull request by YHerm**

### CODE REVIEW

File: src/main/java/edu/greenblitz/tobyDetermined/Nodesssss/MidNode.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import edu.greenblitz.tobyDetermined.subsystems.telescopicArm.Extender.Extender;
+
+import static edu.greenblitz.tobyDetermined.RobotMap.NodeSystem.*;
+
+public class MidNode {
+    private GBNode gbNode;
+    private static MidNode instance;
+
+    private MidNode() {
+        gbNode = null;
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/tobyDetermined/Nodesssss/MidNode.java:14]: why null?

File: src/main/java/edu/greenblitz/tobyDetermined/Nodesssss/MidNode.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        if (start.equals(RobotMap.NodeSystem.SystemsPos.MID_NODE))
+            start = gbNode.getNeighbors().get(0);
+        if (start.toString().contains(systemName1))
+            gbNode = new NodeArm(Extender.getInstance().getLength(), Elbow.getInstance().getAngleRadians());
+        else
+            gbNode = new GriperNode();
+        gbNode.addNeighbors(new RobotMap.NodeSystem.SystemsPos[]{start, end});
+    }
+
+    public GBNode getNodeArm() {
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/tobyDetermined/Nodesssss/MidNode.java:33]: maybe change name to "getGbNode"

File: src/main/java/edu/greenblitz/tobyDetermined/Nodesssss/MultiNodeSubsystem/AStarVertex.java [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+
+import java.util.HashMap;
+import java.util.LinkedList;
+
+import static edu.greenblitz.tobyDetermined.RobotMap.NodeSystem.SystemsPos;
+
+public class AStarVertex {
+
+    public static <T> void printPath(LinkedList<T> pathList) {
+        for (int i = 0; i <= pathList.size() - 1; i++) {
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/tobyDetermined/Nodesssss/MultiNodeSubsystem/AStarVertex.java:14]: same with the for below

```suggestion
        for (int i = 0; i < pathList.size(); i++) {
```

File: src/main/java/edu/greenblitz/tobyDetermined/Nodesssss/MultiNodeSubsystem/AStarVertex.java [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+            System.out.print(pathList.get(i) + ", ");
+        }
+        System.out.println();
+    }
+
+    public static void printPathVer(LinkedList<Vertex> pathList) {
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/tobyDetermined/Nodesssss/MultiNodeSubsystem/AStarVertex.java:20]: does writing "vertex" really hurt?

File: src/main/java/edu/greenblitz/tobyDetermined/commands/ArmCommand.java
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/tobyDetermined/commands/ArmCommand.java:1]: delete file

File: src/main/java/edu/greenblitz/tobyDetermined/Nodesssss/MultiNodeSubsystem/AStarVertex.java [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+            System.out.print(pathList.get(i).getPos1() + ", " + pathList.get(i).getPos2() + ";;;;;");
+        }
+        System.out.println();
+    }
+
+    private static <T> boolean isInList(T object, LinkedList<T> list) {
+        return list.contains(object);
+    }
+
+    private static boolean ifNotInVerList(LinkedList<Vertex> verList, Vertex ver) {
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/tobyDetermined/Nodesssss/MultiNodeSubsystem/AStarVertex.java:31]: just write vertex its 3 letters more, and no need for the if😍​

File: src/main/java/edu/greenblitz/tobyDetermined/Nodesssss/MultiNodeSubsystem/AStarVertex.java [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+        }
+        return true;
+    }
+
+    public static Vertex getVertexLowestFcost(LinkedList<Vertex> open, HashMap<Vertex, Pair<LinkedList<SystemsPos>, Double>> pathMap) {
+        int saveI = 0;
+        int index = 0;
+        double minFCost = Double.MAX_VALUE;
+        while (index != open.size()) {
+            Pair<Boolean, Double> a = returnIsVertexUsableAndCost(open.get(index), pathMap);
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/tobyDetermined/Nodesssss/MultiNodeSubsystem/AStarVertex.java:44]: a like in asshole? or like in Armageddon

File: src/main/java/edu/greenblitz/tobyDetermined/Nodesssss/MultiNodeSubsystem/AStarVertex.java [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+            }
+        }
+        if (isTherePossiblePosition) {
+            return NodeBase.getNode(vertex.getOtherSystemPos()).getOtherSystemMustBeToOut().contains(vertex.getPos1())
+                    || NodeBase.getNode(vertex.getOtherSystemPos()).getOtherSystemMustBeToOut().isEmpty();
+        }
+        return false;
+    }
+
+    public static Double getCostOfOtherSystemPathAndAddToMap(Vertex vertex, HashMap<Vertex, Pair<LinkedList<SystemsPos>, Double>> pathMap) {
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/tobyDetermined/Nodesssss/MultiNodeSubsystem/AStarVertex.java:87]: split into two func

File: src/main/java/edu/greenblitz/tobyDetermined/Nodesssss/MultiNodeSubsystem/AStarVertex.java [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+                }
+            }
+            if (!pathList.get(0).equals(current.getPos1()))
+                pathList.addFirst(current.getPos1());
+            current = parents.get(current);
+        }
+        return new Pair<>(pathList, totalCost);
+    }
+
+    public static Pair<LinkedList<SystemsPos>, Double> getPath(SystemsPos start, SystemsPos end, SystemsPos secondSystemPos) {
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/tobyDetermined/Nodesssss/MultiNodeSubsystem/AStarVertex.java:127]: state the difference between get, return and return final path in the name or get good

File: src/main/java/edu/greenblitz/tobyDetermined/Nodesssss/MultiNodeSubsystem/AStarVertex.java [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+    }
+
+    public static LinkedList<SystemsPos> printAndReturnFinalPath(SystemsPos start, SystemsPos end, SystemsPos secondSystemState) {
+        Pair<LinkedList<SystemsPos>, Double> finalPathAndCost = getPath(start, end, secondSystemState);
+        printPath(finalPathAndCost.getFirst());
+        System.out.println(finalPathAndCost.getSecond());
+        return finalPathAndCost.getFirst();
+    }
+
+    public static void main(String[] args) {
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/tobyDetermined/Nodesssss/MultiNodeSubsystem/AStarVertex.java:166]: the main shouldnt be here

**Review by bossyshrimp68 — COMMENTED** [review]
i dont know if it saves if i leave it open so ill finish my review and continue later🤙​

File: src/main/java/edu/greenblitz/tobyDetermined/Nodesssss/MultiNodeSubsystem/AStarVertex.java (hunk 9 above)
Comment (YHerm) [reply src/main/java/edu/greenblitz/tobyDetermined/Nodesssss/MultiNodeSubsystem/AStarVertex.java:166]: it is

File: src/main/java/edu/greenblitz/tobyDetermined/Nodesssss/NodeBase.java [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+import static edu.greenblitz.tobyDetermined.RobotMap.NodeSystem.SystemsPos.*;
+import static edu.greenblitz.tobyDetermined.RobotMap.NodeSystem.SystemsPos;
+
+public class NodeBase {
+
+    protected final static HashMap<SystemsPos, GBNode> nodeMap = new HashMap<>();
+    private static final LinkedList<SystemsPos> listSystemsPos = new LinkedList<>();
+
+
+    static {
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/tobyDetermined/Nodesssss/NodeBase.java:19]: YOAV

**Review by bossyshrimp68 — CHANGES_REQUESTED** [review]

File: src/main/java/edu/greenblitz/tobyDetermined/subsystems/telescopicArm/Extender/SimulationExtender.java [hunk 11]
```diff
@@ -33,6 +36,7 @@ public void setPower(double power) {
     @Override
     public void setVoltage(double voltage) {
         appliedVoltage = MathUtil.clamp(voltage, -RobotMap.SimulationConstants.MAX_MOTOR_VOLTAGE, RobotMap.SimulationConstants.MAX_MOTOR_VOLTAGE);
+        SmartDashboard.putNumber("simVoltage",appliedVoltage);
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/tobyDetermined/subsystems/telescopicArm/Extender/SimulationExtender.java:39]: remove​❣️​

**Review by bossyshrimp68 — COMMENTED** [review]
good job!! ​😎​

**Pull request closed without merging**